File: assemblyadapter-recycler/src/main/java/com/github/panpf/assemblyadapter/recycler/AssemblyRecyclerAdapter.java
/*
 * Copyright (C) 2017 Peng fei Pan <sky@panpf.me>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.panpf.assemblyadapter.recycler;

import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import com.github.panpf.assemblyadapter.DataManager;
import com.github.panpf.assemblyadapter.Item;
import com.github.panpf.assemblyadapter.ItemFactory;
import com.github.panpf.assemblyadapter.ItemManager;
import com.github.panpf.assemblyadapter.recycler.internal.AssemblyRecyclerItem;

import java.util.Collection;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AssemblyRecyclerAdapter<DATA> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    @NonNull
    private final ItemManager<ItemFactory<?>> itemManager;
    @NonNull
    private final DataManager<DATA> dataManager;

    @Nullable
    private Map<Class<? extends ItemFactory<?>>, ItemSpan> itemSpanMapInGridLayoutManager;
    private boolean notifyOnChange = true;

    @NonNull
    private final DataManager.Callback dataCallback = () -> {
        if (notifyOnChange) {
            notifyDataSetChanged();
        }
    };


    public AssemblyRecyclerAdapter(@NonNull List<ItemFactory<?>> itemFactoryList) {
        this.itemManager = new ItemManager<>(itemFactoryList);
        this.dataManager = new DataManager<>(dataCallback);
    }

    public AssemblyRecyclerAdapter(@NonNull List<ItemFactory<?>> itemFactoryList, @Nullable List<DATA> dataList) {
        this.itemManager = new ItemManager<>(itemFactoryList);
        this.dataManager = new DataManager<>(dataCallback, dataList);
    }

    public AssemblyRecyclerAdapter(@NonNull List<ItemFactory<?>> itemFactoryList, @Nullable DATA[] dataArray) {
        this.itemManager = new ItemManager<>(itemFactoryList);
        this.dataManager = new DataManager<>(dataCallback, dataArray);
    }


    @Override
    public int getItemCount() {
        return dataManager.getDataCount();
    }

    @Nullable
    public DATA getItem(int position) {
        return dataManager.getData(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemViewType(int position) {
        return itemManager.getItemTypeByData(getItem(position));
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemFactory<?> recyclerItemFactory = itemManager.getItemFactoryByItemType(viewType);
        Item<?> item = recyclerItemFactory.dispatchCreateItem(parent);
        AssemblyRecyclerItem<?> recyclerItem = new AssemblyRecyclerItem<>(item);
        applyItemSpanInGridLayoutManager(parent, recyclerItemFactory, recyclerItem);
        return recyclerItem;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int position) {
        if (viewHolder instanceof AssemblyRecyclerItem) {
            //noinspection unchecked
            ((AssemblyRecyclerItem<Object>) viewHolder).dispatchBindData(position, getItem(position));
        }
    }


    @NonNull
    public AssemblyRecyclerAdapter<DATA> setItemSpanInGridLayoutManager(@NonNull Class<? extends ItemFactory<?>> itemFactoryClass, @NonNull ItemSpan itemSpan) {
        if (itemSpanMapInGridLayoutManager == null) {
            itemSpanMapInGridLayoutManager = new HashMap<>();
        }
        itemSpanMapInGridLayoutManager.put(itemFactoryClass, itemSpan);
        if (notifyOnChange) {
            notifyDataSetChanged();
        }
        return this;
    }

    @NonNull
    public AssemblyRecyclerAdapter<DATA> setItemSpanMapInGridLayoutManager(@Nullable Map<Class<? extends ItemFactory<?>>, ItemSpan> itemSpanMap) {
        this.itemSpanMapInGridLayoutManager = itemSpanMap;
        if (notifyOnChange) {
            notifyDataSetChanged();
        }
        return this;
    }

    @Nullable
    public Map<Class<? extends ItemFactory<?>>, ItemSpan> getItemSpanMapInGridLayoutManager() {
        return itemSpanMapInGridLayoutManager;
    }

    private void applyItemSpanInGridLayoutManager(@NonNull ViewGroup parent,
                                                  @NonNull ItemFactory<?> recyclerItemFactory,
                                                  @NonNull AssemblyRecyclerItem<?> recyclerItem) {
        if (itemSpanMapInGridLayoutManager != null && !itemSpanMapInGridLayoutManager.isEmpty() && parent instanceof RecyclerView) {
            RecyclerView.LayoutManager layoutManager = ((RecyclerView) parent).getLayoutManager();
            //noinspection StatementWithEmptyBody
            if (layoutManager instanceof AssemblyGridLayoutManager) {
                // No need to do
            } else if (layoutManager instanceof AssemblyStaggeredGridLayoutManager) {
                ItemSpan itemSpan = itemSpanMapInGridLayoutManager.get(recyclerItemFactory.getClass());
                if (itemSpan != null && itemSpan.span < 0) {
                    View itemView = recyclerItem.getItemView();
                    ViewGroup.LayoutParams layoutParams = itemView.getLayoutParams();
                    if (layoutParams instanceof StaggeredGridLayoutManager.LayoutParams) {
                        ((StaggeredGridLayoutManager.LayoutParams) layoutParams).setFullSpan(true);
                        itemView.setLayoutParams(layoutParams);
                    }
                }
            } else {
                throw new IllegalArgumentException("Since itemSpan is set, the layoutManager of RecyclerView must be AssemblyGridLayoutManager or AssemblyStaggeredGridLayoutManager");
            }
        }
    }


    @NonNull
    public List<DATA> getDataListSnapshot() {
        return dataManager.getDataListSnapshot();
    }

    public void setDataList(@Nullable List<DATA> datas) {
        dataManager.setDataList(datas);
    }

    public boolean addData(@Nullable DATA data) {
        return dataManager.addData(data);
    }

    public void addData(int index, @Nullable DATA data) {
        dataManager.addData(index, data);
    }

    public boolean addAllData(@Nullable Collection<DATA> datas) {
        return dataManager.addAllData(datas);
    }

    @SafeVarargs
    public final boolean addAllData(@Nullable DATA... datas) {
        return dataManager.addAllData(datas);
    }

    public boolean removeData(@Nullable DATA data) {
        return dataManager.removeData(data);
    }

    public DATA removeData(int index) {
        return dataManager.removeData(index);
    }

    public boolean removeAllData(@NonNull Collection<DATA> datas) {
        return dataManager.removeAllData(datas);
    }

    public void clearData() {
        dataManager.clearData();
    }

    public void sortData(@NonNull Comparator<DATA> comparator) {
        dataManager.sortData(comparator);
    }


    public boolean isNotifyOnChange() {
        return notifyOnChange;
    }

    public void setNotifyOnChange(boolean notifyOnChange) {
        this.notifyOnChange = notifyOnChange;
    }


    @NonNull
    public ItemFactory<?> getItemFactoryByItemType(int itemType) {
        return itemManager.getItemFactoryByItemType(itemType);
    }

    @NonNull
    public ItemFactory<?> getItemFactoryByPosition(int position) {
        return getItemFactoryByItemType(itemManager.getItemTypeByData(getItem(position)));
    }
}

File: assemblyadapter-recycler/src/main/java/com/github/panpf/assemblyadapter/recycler/AssemblyRecyclerAdapter.kt
/*
 * Copyright (C) 2021 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.assemblyadapter.recycler

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.github.panpf.assemblyadapter.*
import com.github.panpf.assemblyadapter.internal.ItemDataStorage
import com.github.panpf.assemblyadapter.internal.ItemFactoryStorage
import com.github.panpf.assemblyadapter.recycler.internal.FullSpanSupport
import com.github.panpf.assemblyadapter.recycler.internal.RecyclerViewHolderWrapper

/**
 * An implementation of [RecyclerView.Adapter], which implements multi-type adapters through standardized [ItemFactory].
 * [AssemblyRecyclerAdapter] will use the data corresponding to position to find a matching [ItemFactory] (cannot find an exception will be thrown),
 * and then use [ItemFactory] to create an item view and bind the data
 *
 * @param itemFactoryList The collection of [ItemFactory] passed in from outside, cannot be empty.
 * Each type of data in the data set must have a matching [ItemFactory], otherwise an exception will be thrown
 * @param initDataList Initial data set
 * @see ItemFactory
 */
open class AssemblyRecyclerAdapter<DATA>(
    itemFactoryList: List<ItemFactory<out Any>>,
    initDataList: List<DATA>? = null,
) : RecyclerView.Adapter<RecyclerView.ViewHolder>(), AssemblyAdapter<DATA, ItemFactory<out Any>> {

    private val itemFactoryStorage = ItemFactoryStorage(
        itemFactoryList, "ItemFactory", "AssemblyRecyclerAdapter", "itemFactoryList"
    )
    private val itemDataStorage = ItemDataStorage(initDataList) { oldList, newList ->
        onDataListChanged(oldList, newList)
    }

    /**
     * Get the current list. If a null list is submitted through [submitList], or no list is submitted, an empty list will be returned.
     * The returned list may not change-changes to the content must be passed through [submitList].
     */
    val currentList: List<DATA>
        get() = itemDataStorage.readOnlyList

    init {
        require(itemFactoryList.isNotEmpty()) { "itemFactoryList Can not be empty" }
    }

    /**
     * Set the new list to be displayed.
     */
    open fun submitList(list: List<DATA>?) {
        itemDataStorage.submitList(list)
    }

    override fun getItemCount(): Int {
        return itemDataStorage.dataCount
    }

    fun getItemData(position: Int): DATA {
        return itemDataStorage.getData(position)
    }

    override fun getItemId(position: Int): Long {
        return if (hasStableIds()) {
            val data = getItemData(position) ?: Placeholder
            if (data is ItemId) data.itemId else data.hashCode().toLong()
        } else {
            RecyclerView.NO_ID
        }
    }

    override fun getItemViewType(position: Int): Int {
        val data = getItemData(position) ?: Placeholder
        return itemFactoryStorage.getItemTypeByData(data)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val itemFactory = itemFactoryStorage.getItemFactoryByItemType(viewType)
        val item = itemFactory.dispatchCreateItem(parent)
        return RecyclerViewHolderWrapper(item).apply {
            val layoutManager =
                (parent.takeIf { it is RecyclerView } as RecyclerView?)?.layoutManager
            if (layoutManager is StaggeredGridLayoutManager && layoutManager is FullSpanSupport) {
                (itemView.layoutParams as StaggeredGridLayoutManager.LayoutParams)
                    .isFullSpan = layoutManager.isFullSpanByItemFactoryClass(itemFactory.javaClass)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is RecyclerViewHolderWrapper<*>) {
            @Suppress("UNCHECKED_CAST")
            val item = holder.wrappedItem as Item<Any>
            val data = getItemData(position) ?: Placeholder
            val absoluteAdapterPosition =
                holder.absoluteAdapterPosition.takeIf { it != -1 } ?: holder.position
            item.dispatchBindData(position, absoluteAdapterPosition, data)
        } else {
            throw IllegalArgumentException("holder must be RecyclerViewHolderWrapper")
        }
    }


    override fun getItemFactoryByPosition(position: Int): ItemFactory<Any> {
        val data = getItemData(position) ?: Placeholder
        return itemFactoryStorage.getItemFactoryByData(data) as ItemFactory<Any>
    }

    override fun getItemFactoryByData(data: DATA): ItemFactory<Any> {
        return itemFactoryStorage.getItemFactoryByData(data ?: Placeholder) as ItemFactory<Any>
    }

    override fun <T : ItemFactory<out Any>> getItemFactoryByClass(itemFactoryClass: Class<T>): T {
        return itemFactoryStorage.getItemFactoryByClass(itemFactoryClass)
    }

    protected open fun onDataListChanged(oldList: List<DATA>, newList: List<DATA>) {
        notifyDataSetChanged()
    }
}